internal/server/spanner/model.go

```diff
@@ -32,17 +41,77 @@ type Edge struct {
 	Types       []string `spanner:"types"`
 }
 
-// StatVarObservation struct represents a single row in the StatVarObservation table.
-type StatVarObservation struct {
-	VariableMeasured  string `spanner:"variable_measured"`
-	ObservationAbout  string `spanner:"observation_about"`
-	ObservationDate   string `spanner:"observation_date"`
-	Value             string `spanner:"value"`
-	Provenance        string `spanner:"provenance"`
-	ObservationPeriod string `spanner:"observation_period"`
-	MeasurementMethod string `spanner:"measurement_method"`
-	Unit              string `spanner:"unit"`
-	ScalingFactor     string `spanner:"scaling_factor"`
+// Observation struct represents a single row in the Observation table.
+type Observation struct {
+	VariableMeasured  string     `spanner:"variable_measured"`
+	ObservationAbout  string     `spanner:"observation_about"`
+	Observations      TimeSeries `spanner:"observations"`
+	Provenance        string     `spanner:"provenance"`
+	ObservationPeriod string     `spanner:"observation_period"`
+	MeasurementMethod string     `spanner:"measurement_method"`
+	Unit              string     `spanner:"unit"`
+	ScalingFactor     string     `spanner:"scaling_factor"`
+	ImportName        string     `spanner:"import_name"`
+	ProvenanceURL     string     `spanner:"provenance_url"`
+}
+
+type DateValue struct {
+	Date  string
+	Value float64
+}
+
+type TimeSeries struct {
+	Observations []*DateValue
+}
+
+// DecodeSpanner decodes the observations field to a TimeSeries value.
+// This is inherited from the spanner Decoder interface to decode from a spanner type to a custom type.
+// Reference: https://cloud.google.com/go/docs/reference/cloud.google.com/go/spanner/latest#cloud_google_com_go_spanner_Decoder
+// Note that the undecoded values are of type ListValue and each element a string value.
+func (ts *TimeSeries) DecodeSpanner(val interface{}) (err error) {
+	listVal, ok := val.(*structpb.ListValue)
+	if !ok {
+		return fmt.Errorf("failed to decode TimeSeries: (%v)", val)
+	}
+	for _, v := range listVal.Values {
+		var data map[string]string
+		err := json.Unmarshal([]byte(v.GetStringValue()), &data)
+		if err != nil {
+			return fmt.Errorf("failed to decode TimeSeries value: (%v)", v)
+		}
+		for date, strVal := range data {
+			floatVal, err := strconv.ParseFloat(strVal, 64)
+			if err != nil {
+				return fmt.Errorf("failed to decode TimeSeries float value: (%v)", floatVal)
+			}
+			ts.Observations = append(ts.Observations, &DateValue{
+				Date:  date,
+				Value: floatVal,
+			})
+		}
+	}
+	return nil
+}
+
+// FilterByDate filters observations based on the given date.
+func (ts *TimeSeries) FilterByDate(date string) {
```

Comment:
Just to clarify, are you still planning on caching the entire series? If so, maybe the date filtering be done separately from fetching the full series in GetObservations. Or if not, maybe we can only fetch the requested dates. Wdyt?

Reply:
Good point! Yes, caching eventually. Removed filtering logic from this layer. Will apply in the datasource layer in the next PR.

Reply:
For clarity, I've kept this method but removed calling it from query.go. Will instead call it from the datasource in the next one.

Reply:
Sounds good! Just make sure the datasource caller can handle the case when there's no date match (so shouldn't return anything)

Reply:
Ack.